### PR TITLE
STM32F1 general cleanup and GD32E230 option write special case

### LINKED ISSUE #921: GD32E230 Option byte write needs to be 32bit sized
I've stumbled across an issue that prevented me from resetting the flash protection option bytes. I've traced the issue, and it appears that target_mem_write16 does not work on GD32E230 to write option bytes.
Quick fix below, but not sure how it would affect other targets.

```
static bool stm32f1_option_write_erased(target *t, uint32_t addr, uint16_t value)
{
	if (value == 0xffff)
		return true;
	/* Erase option bytes instruction */
	target_mem_write32(t, FLASH_CR, FLASH_CR_OPTPG | FLASH_CR_OPTWRE);
	target_mem_write32(t, addr, 0xFFFF0000 | value); // <-- change here
	/* Read FLASH_SR to poll for BSY bit */
	while (target_mem_read32(t, FLASH_SR) & FLASH_SR_BSY)
		if(target_check_error(t))
			return false;

	return true;
}
```

### PR DESCRIPTION
**Pull request by perigoso**
trying to do something about #921 
I am not able to physically test this as I don't have any GD32E230 devices

@walmis do you think you could do some testing for me?

fixes #921

Signed-off-by: Rafael Silva <perigoso@riseup.net>

### CODE REVIEW

File: src/target/stm32f1.c [hunk 1]
```diff
@@ -486,9 +499,14 @@ static bool stm32f1_cmd_option(target *t, int argc, const char **argv)
 	target_mem_write32(t, FLASH_OPTKEYR, KEY1);
 	target_mem_write32(t, FLASH_OPTKEYR, KEY2);
 
-	if ((argc == 2) && !strcmp(argv[1], "erase")) {
+	if ((argc == 2) && strcmp(argv[1], "erase") == 0) {
```
Comment (dragonmux) [inline src/target/stm32f1.c:502]: the ()'s around the argc check are redundant and suppress warnings about accidental assignment, please remove them as part of this change

File: src/target/stm32f1.c [hunk 2]
```diff
@@ -486,9 +499,14 @@ static bool stm32f1_cmd_option(target *t, int argc, const char **argv)
...
 		stm32f1_option_erase(t);
-		stm32f1_option_write_erased(t, FLASH_OBP_RDP, flash_obp_rdp_key);
+		bool width_word = false;
+		if (strcmp(t->driver, "GD32E230") == 0) {
```
Comment (dragonmux) [inline src/target/stm32f1.c:505]: If possible, can we detect this based on, eg, t->idcode? This works but we'd like a cheaper option that uses a bit less Flash and time

File: src/target/stm32f1.c [hunk 3]
```diff
@@ -448,16 +453,24 @@ static bool stm32f1_option_write(target *t, uint32_t addr, uint16_t value)
...
+	bool width_word = false;
+	if (strcmp(t->driver, "GD32E230") == 0) {
+		/* GD32E230 special case, target_mem_write16 does not work */
+		width_word = true;
+	}
+
 	for (i = 0; i < 8; i++)
-		if (!(stm32f1_option_write_erased
-			(t, FLASH_OBP_RDP + i*2,opt_val[i])))
+		if (!(stm32f1_option_write_erased(t, FLASH_OBP_RDP + i * 2, opt_val[i], width_word)))
```
Comment (dragonmux) [inline src/target/stm32f1.c:471]: There's an extra pair of redundant brackets here, please remove them

File: src/target/stm32f1.c [hunk 4]
```diff
@@ -448,16 +453,24 @@ static bool stm32f1_option_write(target *t, uint32_t addr, uint16_t value)
...
+
 	/* Check for erased value */
 	if (opt_val[index] != 0xffff)
 		if (!(stm32f1_option_erase(t)))
 			return false;
 	opt_val[index] = value;
+
 	/* Write changed values*/
+	bool width_word = false;
+	if (strcmp(t->driver, "GD32E230") == 0) {
```
Comment (dragonmux) [inline src/target/stm32f1.c:465]: This is covered by the note below about strcmp() vs testing t->idcode or similar

File: src/target/stm32f1.c [hunk 5]
```diff
@@ -418,13 +418,17 @@ static bool stm32f1_option_erase(target *t)
...
 
-static bool stm32f1_option_write_erased(target *t, uint32_t addr, uint16_t value)
+static bool stm32f1_option_write_erased(target *t, uint32_t addr, uint16_t value, bool width_word)
 {
 	if (value == 0xffff)
 		return true;
 	/* Erase option bytes instruction */
 	target_mem_write32(t, FLASH_CR, FLASH_CR_OPTPG | FLASH_CR_OPTWRE);
-	target_mem_write16(t, addr, value);
+	if (width_word) {
```
Comment (dragonmux) [inline src/target/stm32f1.c:427]: the {}'s in this are redundant and not part of the code style. Please remove them

File: src/target/stm32f1.c [hunk 6]
```diff
@@ -440,6 +444,7 @@ static bool stm32f1_option_write(target *t, uint32_t addr, uint16_t value)
 	index = (addr - FLASH_OBP_RDP) / 2;
 	if ((index < 0) || (index > 7))
```
Comment (dragonmux) [inline src/target/stm32f1.c:445]: the ()'s around the individual sub-clauses here are redundant and suppress warnings, please remove them

File: src/target/stm32f1.c [hunk 7]
```diff
@@ -440,6 +444,7 @@ static bool stm32f1_option_write(target *t, uint32_t addr, uint16_t value)
...
 		 return false;
+
 	/* Retrieve old values */
 	for (i = 0; i < 16; i = i +4) {
```
Comment (dragonmux) [inline src/target/stm32f1.c:449]: This feels like i is defined elsewhere and whoever wrote this doesn't know about += - please rewrite this using size_t for i and `i += 4`

File: src/target/stm32f1.c [hunk 8]
```diff
@@ -448,16 +453,24 @@ static bool stm32f1_option_write(target *t, uint32_t addr, uint16_t value)
...
 	opt_val[index] = value;
+
 	/* Write changed values*/
+	bool width_word = false;
+	if (strcmp(t->driver, "GD32E230") == 0) {
+		/* GD32E230 special case, target_mem_write16 does not work */
+		width_word = true;
+	}
+
 	for (i = 0; i < 8; i++)
```
Comment (dragonmux) [inline src/target/stm32f1.c:470]: same comment as above about i defined elsewhere and size_t

**Review by dragonmux — CHANGES_REQUESTED** [review]
This is looking good, though we'd like to see if a cheaper alternative to the strcmp() of `t->driver` is available and, as you're touching this code, clean it up to make it more conformant to the code base style, removing warning-suppressing brackets, etc. (Cleanup should probably be a second commit)

If you don't wish to clean up the code, then no worries, it can be done as a separate pass now we know it needs work.

**Comment by perigoso** [comment]
@dragonmux  theres a code snippet weirding me out 

```
if (0 && flash_obp_rdp_key == FLASH_OBP_RDP_KEY_F3) {
    /* Reload option bytes on F0 and F3*/
    val = adiv5_ap_mem_read(ap, FLASH_CR);
    val |= FLASH_CR_OBL_LAUNCH;
    stm32f1_option_write(t, FLASH_CR, val);
    val &= ~FLASH_CR_OBL_LAUNCH;
    stm32f1_option_write(t, FLASH_CR, val);
}
```
this code block is never executed, I can only assume it's intentional, but i would like to know why, it was added by @UweBonnes  in b1a942aca3e80aaaa8405d313424f1d6f9273bd9

**Comment by dragonmux** [comment]
That's a really fantastic question.. given how old that commit is.. we're going to say just nix it. If we find we need the code we can always pull it back out the Git history.

**Comment by perigoso** [comment]
this should be close to or mergeable, waiting on feedback
there's a couple of commits by now, I can squash some, but I think they're reasonably organized

**Comment by perigoso** [comment]
ready for review

File: src/target/stm32f1.c [hunk 9]
```diff
@@ -253,83 +265,91 @@ static int stm32f1_flash_unlock(target *t, uint32_t bank_offset)
...
 			len = 0;
+
 		addr += f->blocksize;
 	}
 
 	/* Check for error */
 	if (start < FLASH_BANK_SPLIT) {
 		uint32_t sr = target_mem_read32(t, FLASH_SR);
-		if ((sr & SR_ERROR_MASK) || !(sr & SR_EOP)) {
+		if (sr & SR_ERROR_MASK || !(sr & SR_EOP)) {
```
Comment (dragonmux) [inline src/target/stm32f1.c:317]: This pair of ()'s is required as & binds looser than ||

File: src/target/stm32f1.c [hunk 10]
```diff
@@ -253,83 +265,91 @@ static int stm32f1_flash_unlock(target *t, uint32_t bank_offset)
...
 			DEBUG_INFO("stm32f1 flash erase error 0x%" PRIx32 "\n", sr);
 			return -1;
 		}
 	}
-	if ((t->idcode == 0x430) && (end >= FLASH_BANK_SPLIT)) {
+	if (t->idcode == 0x430 && end >= FLASH_BANK_SPLIT) {
 		uint32_t sr = target_mem_read32(t, FLASH_SR + FLASH_BANK2_OFFSET);
-		if ((sr & SR_ERROR_MASK) || !(sr & SR_EOP)) {
+		if (sr & SR_ERROR_MASK || !(sr & SR_EOP)) {
```
Comment (dragonmux) [inline src/target/stm32f1.c:324]: This pair of ()'s is required as & binds looser than ||

**Review by dragonmux — CHANGES_REQUESTED** [review]
This introduces a couple of accidental bugs outlined below, but otherwise looks ready for merge. Good work perigoso!

**Pull request merged**